pkg/apiclient/portforwarder.go

```diff
@@ -0,0 +1,95 @@
+package apiclient
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"log"
+	"net"
+	"net/http"
+	"os"
+
+	"github.com/argoproj/argo-cd/util"
+
+	"github.com/pkg/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/client-go/transport/spdy"
+)
+
+func portForward(podSelector string, namespace string) (int, error) {
+	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+	loadingRules.DefaultClientConfig = &clientcmd.DefaultClientConfig
+	overrides := clientcmd.ConfigOverrides{}
+	clientConfig := clientcmd.NewInteractiveDeferredLoadingClientConfig(loadingRules, &overrides, os.Stdin)
+	config, err := clientConfig.ClientConfig()
+	if err != nil {
+		return -1, err
+	}
+
+	if namespace == "" {
+		namespace, _, err = clientConfig.Namespace()
+		if err != nil {
+			return -1, err
+		}
+	}
+
+	clientSet, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return -1, err
+	}
+
+	pods, err := clientSet.CoreV1().Pods(namespace).List(v1.ListOptions{
+		LabelSelector: podSelector,
+	})
+	if err != nil {
+		return -1, err
+	}
+
+	if len(pods.Items) == 0 {
+		return -1, fmt.Errorf("cannot find argocd-server pod")
+	}
+
+	url := clientSet.CoreV1().RESTClient().Post().
+		Resource("pods").
+		Namespace(pods.Items[0].Namespace).
+		Name(pods.Items[0].Name).
+		SubResource("portforward").URL()
+
+	transport, upgrader, err := spdy.RoundTripperFor(config)
+	if err != nil {
+		return -1, errors.Wrap(err, "Could not create round tripper")
+	}
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, "POST", url)
+
+	readyChan := make(chan struct{}, 1)
+	out := new(bytes.Buffer)
+	errOut := new(bytes.Buffer)
+
+	ln, err := net.Listen("tcp", "[::]:0")
```

Comment:
Hm, maybe the local part of the listener should be configurable for a given host IP, instead of binding to all interfaces? Maybe by using new `--port-forward-addr` switch, which defaults to all addresses/interfaces?

Reply:
This listener is used just to get a random free port. It is immediately closed at line 76. @jannfis is it still might be a security issue to listen to all interfaces ?